metadata_backend/api/parser.py

```diff
@@ -0,0 +1,239 @@
+        except (SchemaNotFoundException, XMLSchemaException) as error:
+            reason = f"{error} {xml_type}"
+            raise web.HTTPBadRequest(reason=reason)
+        return schema
+
+    @staticmethod
+    def sort_actions_by_schemas(data: List) -> List:
+        """Sort schemas according to their priority.
+
+        Schemas need to be processed in certain order (for example 'study'
+        submission needs to processed first in order to generate accession
+        number for other submissions.
+
+        Sorting order is based on ENA Metadata Model:
+        https://ena-docs.readthedocs.io/en/latest/submit/general-guide/metadata.html
+
+        This should probably be refactored later, since submissions can be also
+        made via front-end or with actions in POST-request (i.e. without
+        submission.xml).
+
+        :param data: Data to be sorted
+        :returns: Sorted list
+        """
+        order = {"study": 1, "sample": 2, "experiment": 3, "run": 4,
+                 "analysis": 5, "dac": 6, "policy": 7, "dataset": 8,
+                 "project": 9}
+        return sorted(data, key=lambda x: order[x["schema"]])
+
+    @staticmethod
+    def flatten_and_parse_json(data):
+        """Recursively flatten json and parse away unwanted characters.
+
+        :param data: JSON to be flattened
+        :returns: Flattened JSON
+        """
+        out = {}
+
+        def flatten(content, name=""):
+            if type(content) is dict:
+                for key in content:
+                    flatten(content[key], name + key + "_")
+            elif type(content) is list:
+                i = 0
+                for element in content:
+                    flatten(element, name + str(i) + "_")
+                    i += 1
+            else:
+                if name[0] == "@":
+                    name = name[1:]
+                out[name[:-1]] = content
+
+        flatten(data)
+        return out
+
+    @staticmethod
+    def validate(content: str, schema: XMLSchema) -> None:
+        """Validate XML with XMLSchema instance.
+
+        :param content: XML to be validated
+        :param schema: XMLSchema instance that validates XML.
+
+        :raises: HTTPBadRequest if error was raised during validation
+        """
+        try:
+            schema.validate(content)
+        except (ValueError, XMLSchemaException) as error:
+            raise web.HTTPBadRequest(reason=error)
+
+    def parse_study(self, data: Dict) -> Dict:
+        """Parse data from study-type XML.
+
+        Currently just flattens JSON and generates random accession number as
+        placeholder (numbers are later on fetched from an another source).
+
+        :param xml_content: XML content as JSON
+        :returns: Parsed data as JSON
+        """
+        flattened_data = self.flatten_and_parse_json(data)
+        if "accession" not in flattened_data:
+            flattened_data["accession"] = random.randint(1, 10000)
```

Comment:
usually takes the form of `EGAD51775617678681` EGAD + some nb. for datasets or EGAF + some nb.
we can use  `'EGAD'+''.join(secrets.choice(string.digits) for i in range(16))` to be a a bit more close to what we will have.

`import secrets` is needed

see: https://docs.python.org/3/library/secrets.html